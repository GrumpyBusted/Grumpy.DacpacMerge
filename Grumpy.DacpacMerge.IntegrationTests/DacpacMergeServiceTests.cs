using System;
using System.IO;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using Grumpy.DacpacMerge.IntegrationTests.Helpers;
using Microsoft.SqlServer.Dac;
using Xunit;

namespace Grumpy.DacpacMerge.IntegrationTests
{
    [UseApprovalSubdirectory("Approvals")]
    [UseReporter(typeof(DiffReporter))]
    public sealed class DacpacMergeServiceTests : IDisposable
    {
        private readonly string _databaseName;

        public DacpacMergeServiceTests()
        {
            _databaseName = $"Test_Database_{Guid.NewGuid().ToString()}";
        }

        public void Dispose()
        {
            try
            {
                DatabaseHelper.DropDatabase(_databaseName);
            }
            catch (Exception)
            {
                // ignored
            }
        }

        [Fact]
        public void ApplyToNonExistingDatabase_ShouldThrowException()
        {
            Assert.Throws<DacServicesException>(() => TestRunnerHelper.ExecuteTest("Database3", "SchemaA", _databaseName));
        }

        [Fact]
        public void ApplyNewSchemaToEmptyDatabase_ShouldCreateNewSchema()
        {
            DeployDacpac("Database1");

            var script = TestRunnerHelper.ExecuteTest("Database3", "SchemaB", _databaseName);

            Approvals.Verify(script);
        }

        [Fact]
        public void ApplySameDefinition_ShouldNotAlterDatabase()
        {
            DeployDacpac("Database2");

            var script = TestRunnerHelper.ExecuteTest("Database2", "SchemaA", _databaseName);

            Approvals.Verify(script);
        }

        [Fact]
        public void ApplyNewSchemaToExistingDatabase_ShouldCreateNewSchema()
        {
            DeployDacpac("Database2");

            var script = TestRunnerHelper.ExecuteTest("Database3", "SchemaB", _databaseName);

            Approvals.Verify(script);
        }

        [Fact]
        public void ApplyWithChangedTable_ShouldAlterTable()
        {
            DeployDacpac("Database2");

            var script = TestRunnerHelper.ExecuteTest("Database4", "SchemaA", _databaseName);

            Approvals.Verify(script);
        }

        [Fact]
        public void ApplyWithPostDeploymentScript_ShouldAddPostDeployScript()
        {
            DeployDacpac("Database2");

            var script = TestRunnerHelper.ExecuteTest("Database5", "SchemaA", _databaseName);

            Approvals.Verify(script);
        }

        [Fact]
        public void ApplyWithFunction_ShouldAddFunction()
        {
            DeployDacpac("Database2");

            var script = TestRunnerHelper.ExecuteTest("Database7", "SchemaA", _databaseName);

            Approvals.Verify(script);
        }

        [Fact]
        public void ApplyWithRole_ShouldAddRole()
        {
            DeployDacpac("Database2");

            var script = TestRunnerHelper.ExecuteTest("Database8", "SchemaA", _databaseName);

            Approvals.Verify(script);
        }

        [Fact]
        public void ApplyMultipleSchemas_ShouldChangeBothSchemas()
        {
            DeployDacpac("Database2");

            var script = TestRunnerHelper.ExecuteTest("Database6", "SchemaA;SchemaC", _databaseName);

            Approvals.Verify(script);
        }

        [Fact]
        public void ApplyAllSchemas_ShouldChangeBothSchemas()
        {
            DeployDacpac("Database2");

            var script = TestRunnerHelper.ExecuteTest("Database6", "", _databaseName);

            Approvals.Verify(script);
        }

        [Fact]
        public void ApplyWithChangedItemOutsideSchema_ShouldChangeItem()
        {
            DeployDacpac("Database2");

            var script = TestRunnerHelper.ExecuteTest("Database9", "SchemaA", _databaseName);

            Approvals.Verify(script);
        }

        private void DeployDacpac(string testDatabase)
        {
            var dacpacFile = string.Format(TestRunnerHelper.DacpacLocation, testDatabase);

            using (var package = DacPackage.Load(dacpacFile, DacSchemaModelStorageType.Memory, FileAccess.Read))
            {
                var service = new DacServices(DatabaseHelper.ConnectionString(_databaseName));

                service.Deploy(package, _databaseName);
            }
        }
    }
}
