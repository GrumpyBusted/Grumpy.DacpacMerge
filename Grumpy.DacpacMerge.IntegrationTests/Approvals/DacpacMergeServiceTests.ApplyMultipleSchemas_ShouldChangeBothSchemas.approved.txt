/*
Deployment script for {DatabaseName}

This code was generated by a tool.
Changes to this file may cause incorrect behavior and will be lost if
the code is regenerated.
*/

GO
SET ANSI_NULLS, ANSI_PADDING, ANSI_WARNINGS, ARITHABORT, CONCAT_NULL_YIELDS_NULL, QUOTED_IDENTIFIER ON;

SET NUMERIC_ROUNDABORT OFF;


GO
:setvar DatabaseName "{DatabaseName}"
:setvar DefaultFilePrefix "{DatabaseName}"

GO
:on error exit
GO
/*
Detect SQLCMD mode and disable script execution if SQLCMD mode is not supported.
To re-enable the script after enabling SQLCMD mode, execute the following:
SET NOEXEC OFF; 
*/
:setvar __IsSqlCmdEnabled "True"
GO
IF N'$(__IsSqlCmdEnabled)' NOT LIKE N'True'
    BEGIN
        PRINT N'SQLCMD mode must be enabled to successfully execute this script.';
        SET NOEXEC ON;
    END


GO
USE [$(DatabaseName)];


GO
PRINT N'Creating [SchemaC]...';


GO
CREATE SCHEMA [SchemaC]
    AUTHORIZATION [TestUser];


GO
PRINT N'Altering [SchemaA].[Person]...';


GO
ALTER TABLE [SchemaA].[Person]
    ADD [Birthday] DATETIMEOFFSET (7) NULL;


GO
PRINT N'Creating [SchemaC].[ZipCode]...';


GO
CREATE TABLE [SchemaC].[ZipCode] (
    [Id]      INT           NOT NULL,
    [ZipCode] NVARCHAR (10) NOT NULL,
    [ZipName] NVARCHAR (50) NOT NULL,
    PRIMARY KEY CLUSTERED ([Id] ASC)
);


GO
PRINT N'Creating [SchemaC].[ZipCode].[IX_ZipCode_Name]...';


GO
CREATE NONCLUSTERED INDEX [IX_ZipCode_Name]
    ON [SchemaC].[ZipCode]([ZipCode] ASC);


GO
PRINT N'Update complete.';


GO

